substrate/frame/asset-conversion/ops/src/lib.rs

```diff
@@ -0,0 +1,325 @@
+			// provided. Once the deposit from the prior account is returned, the minted assets will
+			// be burned. Touching the new account is necessary because it's not possible to
+			// transfer assets to the new account if it's required. Additionally, the deposit cannot
+			// be refunded from the prior account until its balance is zero.
+
+			if let Some((d, b)) =
+				T::AssetsRefund::deposit_held(asset1.clone(), prior_account.clone())
+			{
+				let ed = T::DepositAsset::minimum_balance();
+				T::DepositAsset::mint_into(&d, b + ed)?;
+				T::Assets::touch(asset1.clone(), &new_account, &d)?;
+			}
+
+			if let Some((d, b)) =
+				T::AssetsRefund::deposit_held(asset2.clone(), prior_account.clone())
+			{
+				let ed = T::DepositAsset::minimum_balance();
+				T::DepositAsset::mint_into(&d, b + ed)?;
+				T::Assets::touch(asset2.clone(), &new_account, &d)?;
+			}
+
+			if let Some((d, b)) =
+				T::PoolAssetsRefund::deposit_held(info.lp_token.clone(), prior_account.clone())
+			{
+				let ed = T::DepositAsset::minimum_balance();
+				T::DepositAsset::mint_into(&d, b + ed)?;
+				T::PoolAssets::touch(info.lp_token.clone(), &new_account, &d)?;
+			}
+
+			// Transfer all pool related assets to the new account.
+
+			ensure!(
+				balance1 ==
+					T::Assets::transfer(
+						asset1.clone(),
+						&prior_account,
+						&new_account,
+						balance1,
+						Preservation::Expendable,
+					)?,
+				Error::<T>::PartialTransfer
+			);
+
+			ensure!(
+				balance2 ==
+					T::Assets::transfer(
+						asset2.clone(),
+						&prior_account,
+						&new_account,
+						balance2,
+						Preservation::Expendable,
+					)?,
+				Error::<T>::PartialTransfer
+			);
+
+			ensure!(
+				balance3 ==
+					T::PoolAssets::transfer(
+						info.lp_token.clone(),
+						&prior_account,
+						&new_account,
+						balance3,
+						Preservation::Expendable,
+					)?,
+				Error::<T>::PartialTransfer
+			);
+
+			// Refund deposits from prior accounts and burn previously minted assets.
+
+			if let Some((d, b)) =
+				T::AssetsRefund::deposit_held(asset1.clone(), prior_account.clone())
+			{
+				T::AssetsRefund::refund(asset1.clone(), prior_account.clone())?;
+				let ed = T::DepositAsset::minimum_balance();
+				T::DepositAsset::burn_from(&d, b + ed, Precision::Exact, Fortitude::Force)?;
+			}
+
+			if let Some((d, b)) =
+				T::AssetsRefund::deposit_held(asset2.clone(), prior_account.clone())
+			{
+				T::AssetsRefund::refund(asset2.clone(), prior_account.clone())?;
+				let ed = T::DepositAsset::minimum_balance();
+				T::DepositAsset::burn_from(&d, b + ed, Precision::Exact, Fortitude::Force)?;
+			}
+
+			if let Some((d, b)) =
+				T::PoolAssetsRefund::deposit_held(info.lp_token.clone(), prior_account.clone())
+			{
+				T::PoolAssetsRefund::refund(info.lp_token.clone(), prior_account.clone())?;
+				let ed = T::DepositAsset::minimum_balance();
+				T::DepositAsset::burn_from(&d, b + ed, Precision::Exact, Fortitude::Force)?;
+			}
```

Comment:
I would refactor as follows to reuse `ed` and avoid duplicating the `if` statements:
```rust
let ed = T::DepositAsset::minimum_balance();

if let Some((d, b)) =
	T::AssetsRefund::deposit_held(asset1.clone(), prior_account.clone())
{
	T::DepositAsset::mint_into(&d, b.clone() + ed.clone())?;
	T::Assets::touch(asset1.clone(), &new_account, &d)?;
	
	// Transfer all pool related assets to the new account.
	ensure!(
		balance1 ==
			T::Assets::transfer(
				asset1.clone(),
				&prior_account,
				&new_account,
				balance1,
				Preservation::Expendable,
			)?,
		Error::<T>::PartialTransfer
	);
	
	// Refund deposits from prior accounts and burn previously minted assets.
	T::AssetsRefund::refund(asset1.clone(), prior_account.clone())?;
	T::DepositAsset::burn_from(&d, b + ed.clone(), Precision::Exact, Fortitude::Force)?;
}

// Repeat for the remaining assets...
```

Reply:
I don't think it's possible because you want to transfer any deposits that are held, but the assets might exist without deposits (e.g. they are sufficient). Regardless of the deposits though, you want to do `T::Assets::transfer(...)`, which is the main objective of the extrinsic.

Most you could do is cache `T::AssetsRefund::deposit_held(asset1.clone(), prior_account.clone())` to some `maybe_asset_1_deposit` and query it later to avoid making the call again as they should be unchanged by that point in the code.

Reply:
Right, I was wrongly presuming all assets would have deposits.